File: wallendar-app/app/src/main/java/com/td/wallendar/di/ProductionDependenciesContainer.java
package com.td.wallendar.di;

import android.content.Context;
import android.content.SharedPreferences;

import com.td.wallendar.repositories.interfaces.ChargesRepository;
import com.td.wallendar.repositories.interfaces.DebtsRepository;
import com.td.wallendar.repositories.interfaces.GroupsRepository;
import com.td.wallendar.service.ChargesService;
import com.td.wallendar.service.DebtsService;
import com.td.wallendar.service.GroupsService;
import com.td.wallendar.utils.scheduler.SchedulerProvider;

import retrofit2.Retrofit;

public class ProductionDependenciesContainer implements DependenciesContainer {

    private final DependenciesModule dependenciesModule;

    private SchedulerProvider schedulerProvider;
    private SharedPreferences loginSharedPreferences;
    private Retrofit retrofit;
    private GroupsRepository groupsRepository;
    private ChargesRepository chargesRepository;
    private DebtsRepository debtsRepository;
    private GroupsService groupsService;
    private ChargesService chargesService;
    private DebtsService debtsService;

    public ProductionDependenciesContainer(final Context context) {
        dependenciesModule = new DependenciesModule(context);
    }


    @Override
    public Context getApplicationContext() {
        return dependenciesModule.getApplicationContext();
    }

    @Override
    public SharedPreferences getLoginSharedPreferences() {
        if (loginSharedPreferences == null) {
            loginSharedPreferences = dependenciesModule.provideLoginSharedPreferences();
        }
        return loginSharedPreferences;
    }

    @Override
    public SchedulerProvider getSchedulerProvider() {
        if (schedulerProvider == null) {
            schedulerProvider = dependenciesModule.provideScheduler();
        }
        return schedulerProvider;
    }

    @Override
    public GroupsRepository getGroupsRepository() {
        if (groupsRepository == null) {
            groupsRepository = dependenciesModule.provideGroupsRepository(getGroupsService());
        }
        return groupsRepository;
    }

    @Override
    public ChargesRepository getChargesRepository() {
        if (chargesRepository == null) {
            chargesRepository = dependenciesModule.provideChargesRepository(getChargesService());
        }
        return chargesRepository;
    }

    @Override
    public DebtsRepository getDebtsRepository() {
        if (debtsRepository == null) {
            debtsRepository = dependenciesModule.provideDebtsRepository(getDebtsService());
        }
        return debtsRepository;
    }

    @Override
    public Retrofit getRetrofit() {
        if (retrofit == null) {
            retrofit = dependenciesModule.provideRetrofit();
        }
        return retrofit;
    }

    @Override
    public GroupsService getGroupsService() {
        if (groupsService == null) {
            groupsService = dependenciesModule.provideGroupsService(getRetrofit());
        }
        return groupsService;
    }

    @Override
    public ChargesService getChargesService() {
        if (chargesService == null) {
            chargesService = dependenciesModule.provideChargesService(getRetrofit());
        }
        return chargesService;
    }

    @Override
    public DebtsService getDebtsService() {
        if (debtsService == null) {
            debtsService = dependenciesModule.provideDebtsService(getRetrofit());
        }
        return debtsService;
    }
}

File: wallendar-app/app/src/main/java/com/td/wallendar/di/ProductionDependenciesContainer.kt
package com.td.wallendar.di

import android.content.Context
import android.content.SharedPreferences
import com.td.wallendar.repositories.interfaces.*
import com.td.wallendar.service.*
import com.td.wallendar.utils.scheduler.SchedulerProvider
import retrofit2.Retrofit

class ProductionDependenciesContainer(context: Context) : DependenciesContainer {
    private val dependenciesModule: DependenciesModule = DependenciesModule(context)
    private var schedulerProvider: SchedulerProvider? = null
    private var loginSharedPreferences: SharedPreferences? = null
    private var retrofit: Retrofit? = null
    private var groupsRepository: GroupsRepository? = null
    private var eventsRepository: EventsRepository? = null
    private var chargesRepository: ChargesRepository? = null
    private var debtsRepository: DebtsRepository? = null
    private var applicationUsersRepository: ApplicationUsersRepository? = null
    private var groupsService: GroupsService? = null
    private var eventsService: EventsService? = null
    private var chargesService: ChargesService? = null
    private var debtsService: DebtsService? = null
    private var applicationUsersService: ApplicationUsersService? = null

    override fun getApplicationContext(): Context {
        return dependenciesModule.getApplicationContext()
    }

    override fun getLoginSharedPreferences(): SharedPreferences {
        if (loginSharedPreferences == null) {
            loginSharedPreferences = dependenciesModule.provideLoginSharedPreferences()
        }
        return loginSharedPreferences!!
    }

    override fun getSchedulerProvider(): SchedulerProvider {
        if (schedulerProvider == null) {
            schedulerProvider = dependenciesModule.provideScheduler()
        }
        return schedulerProvider!!
    }

    override fun getGroupsRepository(): GroupsRepository {
        if (groupsRepository == null) {
            groupsRepository = dependenciesModule.provideGroupsRepository(getGroupsService())
        }
        return groupsRepository!!
    }

    override fun getEventsRepository(): EventsRepository {
        if (eventsRepository == null) {
            eventsRepository = dependenciesModule.provideEventsRepository(getEventsService())
        }
        return eventsRepository!!
    }

    override fun getEventsService(): EventsService {
        if (eventsService == null) {
            eventsService = dependenciesModule.provideEventsService(getRetrofit())
        }
        return eventsService!!
    }

    override fun getChargesRepository(): ChargesRepository {
        if (chargesRepository == null) {
            chargesRepository = dependenciesModule.provideChargesRepository(getChargesService())
        }
        return chargesRepository!!
    }

    override fun getDebtsRepository(): DebtsRepository {
        if (debtsRepository == null) {
            debtsRepository = dependenciesModule.provideDebtsRepository(getDebtsService())
        }
        return debtsRepository!!
    }

    override fun getApplicationUsersRepository(): ApplicationUsersRepository {
        if (applicationUsersRepository == null) {
            applicationUsersRepository = dependenciesModule.provideApplicationUsersRepository(getApplicationUsersService())
        }
        return applicationUsersRepository!!
    }

    override fun getRetrofit(): Retrofit {
        if (retrofit == null) {
            retrofit = dependenciesModule.provideRetrofit()
        }
        return retrofit!!
    }

    override fun getGroupsService(): GroupsService {
        if (groupsService == null) {
            groupsService = dependenciesModule.provideGroupsService(getRetrofit())
        }
        return groupsService!!
    }

    override fun getChargesService(): ChargesService {
        if (chargesService == null) {
            chargesService = dependenciesModule.provideChargesService(getRetrofit())
        }
        return chargesService!!
    }

    override fun getDebtsService(): DebtsService {
        if (debtsService == null) {
            debtsService = dependenciesModule.provideDebtsService(getRetrofit())
        }
        return debtsService!!
    }

    override fun getApplicationUsersService(): ApplicationUsersService {
        if (applicationUsersService == null) {
            applicationUsersService = dependenciesModule.provideApplicationUsersService(getRetrofit())
        }
        return applicationUsersService!!
    }

}